src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,603 @@
+                ArrayPool<byte>.Shared.Return(rentedBytes);
+            }
+
+            void WriteCore(ReadOnlySpan<byte> remainingOuterEncodedBytes, char[] scratchChars, byte[] scratchBytes)
+            {
+                bool decoderFinished, encoderFinished;
+                do
+                {
+                    // convert bytes [this] -> chars
+
+                    _thisDecoder.Convert(
+                        bytes: remainingOuterEncodedBytes,
+                        chars: scratchChars,
+                        flush: false,
+                        out int bytesConsumed,
+                        out int charsWritten,
+                        out decoderFinished);
+
+                    remainingOuterEncodedBytes = remainingOuterEncodedBytes[bytesConsumed..];
+
+                    // convert chars -> bytes [inner]
+
+                    Span<char> decodedChars = scratchChars.AsSpan(..charsWritten);
+
+                    do
+                    {
+                        _innerEncoder.Convert(
+                            chars: decodedChars,
+                            bytes: scratchBytes,
+                            flush: false,
+                            out int charsConsumed,
+                            out int bytesWritten,
+                            out encoderFinished);
+
+                        decodedChars = decodedChars[charsConsumed..];
+
+                        // It's more likely that the inner stream provides an optimized implementation of
+                        // Write(byte[], ...) over Write(ROS<byte>), so we'll prefer the byte[]-based overloads.
+
+                        _innerStream.Write(scratchBytes, 0, bytesWritten);
+                    } while (!encoderFinished);
+                } while (!decoderFinished);
+            }
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+        {
+            EnsurePreWriteConditions();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return new ValueTask(Task.FromCanceled(cancellationToken));
+            }
+
+            return WriteAsyncCore(buffer, cancellationToken);
+            async ValueTask WriteAsyncCore(ReadOnlyMemory<byte> remainingOuterEncodedBytes, CancellationToken cancellationToken)
+            {
+                int rentalLength = Math.Clamp(remainingOuterEncodedBytes.Length, MinWriteRentedArraySize, MaxWriteRentedArraySize);
+
+                char[] scratchChars = ArrayPool<char>.Shared.Rent(rentalLength);
```

Comment:
Why does `WriteAsync` rents the arrays inside `WriteAsyncCore` and the sync counterpart `Write` does not?

Reply:
This was an artifact of an earlier iteration that I forgot to clean up.